File: dynamic programming/fibonacci_dynamic_programming/python/fib.py
"""
practice dynamic programming fibonacci function
"""


# dynamic programming fibonacci

def fib(n, memo={1 : 1, 2 : 1}):
    if n <= 2:
        return memo[n]
    else:
        if memo.get(n):
            return memo[n]
        
        else:
            memo[n] = fib(n - 1, memo) + fib(n - 2, memo)
            return memo[n]


print(fib(50))


 

File: dynamic programming/fibonacci_dynamic_programming/c++/fib.cpp

#include <iostream>
#include <vector>

size_t fib(size_t n,std::vector<size_t>&memo)
{
    if (n <= 2)
    {
        return memo[n];
    }
    else
    {
        if (n < memo.size())
            return memo[n];
        memo.push_back( fib(n - 1, memo) + fib(n - 2, memo));
        return memo[n];
    }
}




int main()
{
    std::vector<size_t> mem;
    mem.push_back(1);
    mem.push_back(1);
    mem.push_back(1);
    std::cout << fib(50, mem) << std::endl;
    return 0;
}